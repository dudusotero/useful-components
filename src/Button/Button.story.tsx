import React, { FC } from 'react'
import { boolean, select } from '@storybook/addon-knobs'
import { Button } from './'

export default {
  component: Button,
  title: 'Button',
}

export const Simple: FC = () => (
  <Button
    disabled={boolean('Disabled', false)}
    kind={select('Kind', ['contained', 'outlined', 'text'], 'contained')}
    size={select('Size', ['small', 'medium', 'large'], 'medium')}
    tier={select('Tier', ['primary', 'secondary'], 'primary')}
  >
    Simple
  </Button>
)
