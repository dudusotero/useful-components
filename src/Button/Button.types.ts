import { ButtonHTMLAttributes, ReactNode } from 'react'

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children?: ReactNode
  size?: 'small' | 'medium' | 'large'
  kind?: 'contained' | 'outlined' | 'text'
  tier?: 'primary' | 'secondary'
}

export type ButtonRef = HTMLButtonElement

export const defaultProps: ButtonProps = {
  kind: 'contained',
  size: 'medium',
  tier: 'primary',
}
