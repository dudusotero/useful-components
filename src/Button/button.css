:root {
  --uc-font-size: 1rem;
  --uc-background-color: #ff0000;
}

.uc-button {
  display: inline-block;
  font-size: var(--uc-font-size);
  background-color: var(--uc-background-color);
}
