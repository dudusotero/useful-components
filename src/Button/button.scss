@import '../variables';

.artio-button {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen,
    Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 1.25rem;
  max-width: 100%;
  width: auto;
  padding-top: 0;
  padding-bottom: 0;

  &:focus {
    outline: none;
  }

  &[data-kind='contained'] {
    transition: background-color 0.2s linear, box-shadow 0.2s linear,
      color 0.2s linear;
    background-color: $primary400;
    box-shadow: none;
    border: none;
    gap: 0.6875rem;
    color: $white;

    &:hover {
      background-color: $primary200;

      -webkit-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      -moz-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      box-shadow: 0 0 0.4rem 0.1rem rgba($color: $primary-shadow, $alpha: 0.75);
    }

    &:focus {
      background-color: $primary800;
      box-shadow: none;
    }

    &:active {
      background-color: $primary600;
      -webkit-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      -moz-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      box-shadow: 0 0 0.4rem 0.1rem rgba($color: $primary-shadow, $alpha: 0.75);
    }

    &:disabled {
      background-color: $primary900;
      color: $primary800;
      box-shadow: none;
      cursor: not-allowed;
    }
  }

  &[data-kind='outlined'] {
    transition: background-color 0.2s linear, box-shadow 0.2s linear,
      border-color 0.2s linear, color 0.2s linear;
    background-color: transparent;
    border: 0.125rem solid $primary400;
    box-shadow: none;
    gap: 0.6875rem;
    color: $primary400;

    &:hover {
      background-color: $primary200;
      border-color: $primary100;

      -webkit-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      -moz-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      box-shadow: 0 0 0.4rem 0.1rem rgba($color: $primary-shadow, $alpha: 0.75);
      color: $white;
    }

    &:focus {
      background-color: $primary800;
      border-color: $primary100;
      box-shadow: none;
      color: $primary100;
    }

    &:active {
      background-color: $primary600;
      border-color: $primary400;
      -webkit-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      -moz-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      box-shadow: 0 0 0.4rem 0.1rem rgba($color: $primary-shadow, $alpha: 0.75);
    }

    &:disabled {
      background-color: transparent;
      border-color: $primary900;
      color: $primary900;
      box-shadow: none;
      cursor: not-allowed;
    }
  }

  &[data-kind='text'] {
    transition: background-color 0.2s linear, box-shadow 0.2s linear,
      border-color 0.2s linear, color 0.2s linear;
    background-color: transparent;
    border: none;
    box-shadow: none;
    gap: 0.6875rem;
    color: $primary200;

    &:hover {
      background-color: $primary200;
      -webkit-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      -moz-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      box-shadow: 0 0 0.4rem 0.1rem rgba($color: $primary-shadow, $alpha: 0.75);
      color: $white;
    }

    &:focus {
      background-color: $primary800;
      box-shadow: none;
      color: $primary100;
    }

    &:active {
      background-color: $primary600;
      -webkit-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      -moz-box-shadow: 0 0 0.4rem 0.1rem
        rgba($color: $primary-shadow, $alpha: 0.75);
      box-shadow: 0 0 0.4rem 0.1rem rgba($color: $primary-shadow, $alpha: 0.75);
    }

    &:disabled {
      background-color: transparent;
      color: $primary900;
      box-shadow: none;
      cursor: not-allowed;
    }
  }

  &[data-size='small'] {
    height: 2rem;
    font-size: 0.8rem;
    gap: 0.6rem;
    padding-right: 1rem;
    padding-left: 1rem;
    width: auto;

    i {
      font-size: 1rem;
    }
  }

  &[data-size='medium'] {
    height: 2.5rem;
    font-size: 1rem;
    font-weight: 500;
    padding-right: 1.5rem;
    padding-left: 1.5rem;
    gap: 0.5rem;

    i {
      font-size: 1.5rem;
    }
  }

  &[data-size='large'] {
    height: 3rem;
    font-size: 1.3rem;
    border-radius: 1.5rem;
    font-weight: 500;
    padding-right: 1.5rem;
    padding-left: 1.5rem;

    i {
      font-size: 1.5rem;
    }
  }
}
