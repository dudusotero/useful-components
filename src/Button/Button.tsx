import React, { forwardRef } from 'react'

import { clsx } from '../utils'

import { ButtonProps, ButtonRef, defaultProps } from './Button.types'
import './button.scss'

const Button = forwardRef<ButtonRef, ButtonProps>(
  ({ children, kind, size, tier, ...props }, ref) => {
    return (
      <button
        {...props}
        ref={ref}
        className={clsx('artio-button', props.className)}
        data-kind={kind}
        data-size={size}
        data-tier={tier}
      >
        {children}
      </button>
    )
  }
)

Button.defaultProps = defaultProps

export default Button
