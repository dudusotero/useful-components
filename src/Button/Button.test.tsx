import React from 'react'
import renderer from 'react-test-renderer'
import { Button } from './'
import { defaultProps } from './Button.types'

it('renders correctly', (): void => {
  const tree = renderer.create(<Button />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('joins default class with className prop', (): void => {
  const { root } = renderer.create(<Button className='button' />)

  expect(root.findByType('button').props.className).toEqual(
    'artio-button button'
  )
})

it('check the default values', (): void => {
  const { root } = renderer.create(<Button />)
  const buttonProps = root.findByType('button').props

  expect([
    buttonProps['data-kind'],
    buttonProps['data-size'],
    buttonProps['data-tier'],
  ]).toEqual([defaultProps.kind, defaultProps.size, defaultProps.tier])
})
