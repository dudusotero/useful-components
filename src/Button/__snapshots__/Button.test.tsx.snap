// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="artio-button"
  data-kind="contained"
  data-size="medium"
  data-tier="primary"
/>
`;
